### PR TITLE
Fixes #1903 - Re-use Connections class in SP

### LINKED ISSUE #1903: Re-use Connections class in SpatialPooler class
In order to simplify/uniformizy current spatial and temporal implementations and ease future changes, I think it's very insteresting we implement Connections class in SpatialPooler class. This already was done nicely in htm.java by @cogmission and it seems that works fine. We would move such class into a `connections.py` file and make TemporalMemory and SpatialPooler import it.

Connections class in Java:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/Connections.java

SpatialPooler class in Java re-using connections:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/SpatialPooler.java#L73

TemporalMemory class in Java re-using connections:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/TemporalMemory.java#L67

@chetan51 @cogmission

### PR DESCRIPTION
**Pull request by david-ragazzi**
Fixes: https://github.com/numenta/nupic/issues/1903
Depends: https://github.com/numenta/nupic/issues/1449

Ladies and gentlemen, now SP also has connections! Connections class is a prime by @chetan51 which allow us to separate net specification from net state. This seems simple but has a lot of implications to make nupic easier to understanding and mantaining. This class is so abstract when comes to segments and synapses that I managed adapt it to SP without need create any additional method but only re-using the existing ones.

Some important changes:

- `_potentialPools` was removed: Now SP create synapses to all input bits that are in potential area of a column. So when methods like `_adaptSynapses`, `_bumpUpWeakColumns`, `_updatePermanencesForColumn`, need update synapses's permanencies, they don't need handle a `mapPotential` but simply handle directly such synapses of the column or get the indices of the input bits which they are connected.

- `_permanences` was removed: Now SP uses a dict to get permanencies from an individual column and put them into a temporary numpy array to perform changes (`perm`). After permanencies operations is done, SP updates the synapses of a column with the new permanencies values (using _updatePermanencesForColumn method).

I compared the performance without or with these changes and I feel litle difference. But it is desirable we have more eyes on this.

This PR still is not ready because I still have to implement serialization for Connections and more tests are need to we find areas for improvement. 

TODO:
- [X] Move Connections class from TemporalMemory into a separated file.
- [X] Adapt SP to use Connections.
- [ ] Create serialization for Connections.
- [ ] Compare performance.

### CODE REVIEW

**Comment by breznak** [comment]
wow :cool: A really exciting change! I think I'll need a brain for the review, so I'll leave that for tmr, but it looks really :+1: :+1:

**Comment by david-ragazzi** [comment]
@breznak haha.. Really this won't an easy job, I even had took advantage that I was with fresh brain and without distractions to finally work on this.. If this improvement has similar or better performance (who knows), then we should say goodbye to the sparse matrix.. and so finally SP will be more intuitive to understand.. :-D

Talking about performance, in the current state of this PR, SP has to transverse synapses to get the input indices which they are connected at every time that such indices are need. I'm thinking of use dicts for perform this operation once and so gain performance.. I'm open to suggestions..

**Comment by breznak** [comment]
I think this would also fix this old pal, https://github.com/numenta/nupic/issues/1385 
because the methods are not used at all after this PR

**Comment by breznak** [comment]
I've just cleaned up the python benchmark PR, so you can use https://github.com/numenta/nupic/pull/1928 to measure the new SP performance.

File: nupic/research/spatial_pooler.py [hunk 1]
```diff
@@ -965,9 +964,10 @@ def _adaptSynapses(self, inputVector, activeColumns):
     permChanges.fill(-1 * self._synPermInactiveDec)
     permChanges[inputIndices] = self._synPermActiveInc
     for i in activeColumns:
-      perm = self._permanences.getRow(i)
-      maskPotential = numpy.where(self._potentialPools.getRow(i) > 0)[0]
-      perm[maskPotential] += permChanges[maskPotential]
+      synapses = self.connections.synapsesForSegment(i)
+      inputIndices = self._getInputIndexesFromSynapses(synapses)
+      perm = self._getPermanencesFromColumn(i)
+      perm += permChanges[inputIndices]
```
Comment (breznak) [inline nupic/research/spatial_pooler.py]: some trouble in adaptSynapses() and the failing test https://travis-ci.org/numenta/nupic/jobs/59624260#L3493

File: nupic/research/spatial_pooler.py [hunk 2]
```diff
@@ -1091,19 +1108,16 @@ def _initPermNonConnected(self):
...
+    array the size of the input, where each entry in the array represents the
+    initial permanence value between the input bit at the particular index in
+    the array, and the column represented by the 'index' parameter.
 
     Parameters:
     ----------------------------
-    @param potential: A numpy array specifying the potential pool of the column.
-                    Permanence values will only be generated for input bits
-                    corresponding to indices for which the mask value is 1.
+    @param index:        The index identifying a column.
```
Comment (breznak) [inline nupic/research/spatial_pooler.py]: Some more details how is the index computed? eg for a 2D SP?

File: nupic/research/spatial_pooler.py [hunk 3]
```diff
@@ -1162,16 +1173,37 @@ def _mapColumn(self, index):
...
+  def _createSynapsesForColumn(self, index, wrapAround=False):
+    """
+    Maps a column to its input bits, create synapses from the column to these,
+    and initializes the permanences of these synapses. This method encapsulates
+    the topology of the region. It takes the index of the column as an argument
+    and determines what are the indices of the input vector that are located
+    within the column's potential pool. The return value is a list containing
+    the indices of the input bits. The current implementation of the base class
+    only supports a 1 dimensional topology of columns with a 1 dimensional
+    topology of inputs. To extend this class to support 2-D topology you will
```
Comment (breznak) [inline nupic/research/spatial_pooler.py]: does it mean `Connections` does not support 2D SP columns, while current implementation does?

File: nupic/research/connections.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    self._synapsesForSegment = dict()
+    self._synapsesForPresynapticCell = defaultdict(dict)
+
+    # Index of the next segment to be created
+    self._nextSegmentIdx = 0
+    # Index of the next synapse to be created
+    self._nextSynapseIdx = 0
+
+
+  def elementForSegment(self, segment):
```
Comment (breznak) [inline nupic/research/connections.py]: nit: maybe call this `getElement()`?

File: nupic/research/spatial_pooler.py (hunk 1 above)
Comment (david-ragazzi) [inline nupic/research/spatial_pooler.py]: > some trouble in adaptSynapses() and the failing test https://travis-ci.org/numenta/nupic/jobs/59624260#L3493

Yep, I did fix it.

File: nupic/research/spatial_pooler.py (hunk 2 above)
Comment (david-ragazzi) [inline nupic/research/spatial_pooler.py]: > Some more details how is the index computed? eg for a 2D SP?

Could be. I simply get existing info and add more some details.

File: nupic/research/spatial_pooler.py (hunk 3 above)
Comment (david-ragazzi) [inline nupic/research/spatial_pooler.py]: > does it mean Connections does not support 2D SP columns, while current implementation does?

Actually this description has text that was "stolen" from `_mapPotential` which its desciption is focused in 2D.

File: nupic/research/connections.py (hunk 4 above)
Comment (david-ragazzi) [inline nupic/research/connections.py]: > nit: maybe call this getElement()?

Could be. `elementForSegment` is original call form used by @chetan51 .

**Comment by david-ragazzi** [comment]
This PR is blocked because the "pickes": it doesn't allow Connections class be serialized. Let's wait CapnProto be the official serialiation method to move forward.

In the meantime, I'll create a PR for move Connections class from TemporalMemory and so decrease the files to review.

**Comment by rhyolight** [comment]
Depends on https://github.com/numenta/nupic/issues/1449

**Comment by chetan51** [comment]
@david-ragazzi You can now consolidate this PR with https://github.com/numenta/nupic/pull/2080.

**Comment by david-ragazzi** [comment]
> @david-ragazzi You can now consolidate this PR with #2080.

Yep, when I go home I'll rebase this PR and revert the related commit (https://github.com/david-ragazzi/nupic/commit/b3af7302f2e3c68d533e4136f5b19803167001af).

**Comment by david-ragazzi** [comment]
> @david-ragazzi You can now consolidate this PR with #2080.

Done.

**Comment by cogmission** [comment]
@david-ragazzi  @rhyolight  Was this merged along with #2080 ?

**Comment by david-ragazzi** [comment]
@david-ragazzi @rhyolight Was this merged along with #2080 ?

Nope. #2080 was only to move Connections from TemporalMemory...

File: nupic/research/spatial_pooler.py [hunk 5]
```diff
@@ -1308,7 +1358,9 @@ def _calculateOverlap(self, inputVector):
                     the spatial pooler.
     """
     overlaps = numpy.zeros(self._numColumns).astype(realDType)
-    self._connectedSynapses.rightVecSumAtNZ_fast(inputVector, overlaps)
+    for i in xrange(self._numColumns):
+      inputIndices = self._getInputIndicesFromSynapsesForColumn(i)
```
Comment (david-ragazzi) [inline nupic/research/spatial_pooler.py]: TODO: I need use `onlyConnected=True` here.

**Comment by cogmission** [comment]
Does anyone know the status of this?

**Comment by breznak** [comment]
afaik, blocked by "serialization stuff"

**Comment by cogmission** [comment]
Ah thanks!

**Comment by chetan51** [comment]
Serialization for Connections is completed: https://github.com/numenta/nupic/pull/2100

**Comment by david-ragazzi** [comment]
> Serialization for Connections is completed: #2100

I think this is not enough because models saved with pickle need be re-serialized to capnproto using the script addressed by https://github.com/numenta/nupic/issues/1418.

**Pull request closed without merging**